TilerServer/TilerWebModule.pas

```diff
@@ -3698,69 +3703,518 @@ function TSliceDiffGeometryICLR.GenerateTileCalc(const aExtent: TExtent; aBitmap
+begin
+  if not IsNaN(aActiveValue) then
+  begin
+    if not IsNaN(aRefValue) then
+      Result := (aRefValue + aActiveValue) / 2
+    else Result := aActiveValue;
+  end
+  else
+  begin
+    if not IsNaN(aRefValue)
+    then Result := aRefValue
+    else aValidFlag := False;
+  end;
+end;
+
+function TSliceDiffGeometryICLR.GetPaletteColor(aActiveTexture, aRefTexture: Double; var aValidFlag: Boolean): TGeoColors;
+begin
+  if not (IsNaN(aActiveTexture) or IsNaN(aRefTexture))then
+    Result := fPalette.ValueToColors(aActiveTexture-aRefTexture)
+  else aValidFlag := False;
+end;
+
+function TSliceDiffGeometryICLR.ConstructPolygon(aPolyPoints: TArray<Double>): TPolygon;
+var
+  polygon: TPolygon;
+  ele: Integer;
+begin
+  setLength(polygon, 5);
+  ele := 0;
+  While ele < length(aPolyPoints) do
+  begin
+    polygon[trunc(ele/2)].X := aPolyPoints[ele];
+    inc(ele,1);
+    polygon[trunc(ele/2)].Y := aPolyPoints[ele];
+    inc(ele,1);
+  end;
+
+  polygon[trunc(ele/2)].X := aPolyPoints[0];
+  ele := ele+1;
+  polygon[trunc(ele/2)].Y := aPolyPoints[1];
+
+  Result := polygon;
+end;
+
+procedure TSliceDiffGeometryICLR.DrawFillPolygon(aColor: TGeoColors; aBitmap: FMX.Graphics.TBitmap; aPolygon: TPolygon);
+begin
+  if aColor.fillColor<>0  then
+  begin
+      aBitmap.Canvas.Fill.Color := aColor.fillColor;
+      aBitmap.Canvas.FillPolygon(aPolygon, 1);
+  end;
+  if aColor.outlineColor<>0 then
+  begin
+      aBitmap.Canvas.Stroke.Color := aColor.outlineColor;
+      aBitmap.Canvas.DrawPolygon(aPolygon, 1);
+  end;
+end;
+
+{ TSliceDiffGeometryICLR2 }
+
+constructor TSliceDiffGeometryICLR2.Create(aLayer: TLayer; aPalette: TWDPalette; aTimeStamp: TDateTime; aCurrentSlice, aRefSlice: TSliceGeometryICLR);
+begin
+  inherited Create(aLayer, aPalette, aTimeStamp, aCurrentSlice, aRefSlice);
+end;
+
+destructor TSliceDiffGeometryICLR2.Destroy;
+begin
+  inherited;
+end;
+
+function TSliceDiffGeometryICLR2.GenerateTileCalc(const aExtent: TExtent; aBitmap: FMX.Graphics.TBitmap; aPixelWidth, aPixelHeight: Double): TGenerateTileStatus;
+var
+  isgop: TPair<TWDID, TSliceGeometryICLRObject>;
+  polygonCommon: TPolygon;
+  capacityFactor: Double;
+  // polygon drawing
+  part: TWDGeometryPart;
+  point: TWDGeometryPoint;
+  path: TPathData;
+  xCurr, yCurr, xPrev, yPrev, xn, yn: Double;
+  pointDist: Double;
+  bufferExtent: TExtent;
+  widthL, widthR: Double;
+  colorsL: TGeoColors;
+  colorsR: TGeoColors;
+  validL, validR: Boolean;
+  refObj: TSliceGeometryICLRObject;
+
+  xDCommon, yDCommon, xDExtra, yDExtra: Double;
+  polygonExtra: TPolygon;
+  isCommonPoly: Boolean;
+  polyPoints: TArray<Double>;
+begin
+  widthL := Double.NaN;
+  widthR := Double.NaN;
+  Result := gtsFailed; // sentinel
+  if Assigned(fPalette) then
+  begin
+    aBitmap.Canvas.BeginScene;
+    try
+      aBitmap.Canvas.Clear(0);
+      aBitmap.Canvas.Fill.Kind := TBrushKind.Solid;
+      capacityFactor := 0.001/Abs(aExtent.YMax-aExtent.YMin);

```

Comment:
Please add a check that ensures YMax and YMin are different.

Reply:
Done!